pkg/operator/controller/route-metrics/controller.go

```diff
@@ -0,0 +1,178 @@
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	routev1 "github.com/openshift/api/route/v1"
+)
+
+const (
+	controllerName = "route_metrics_controller"
+)
+
+var (
+	log = logf.Logger.WithName(controllerName)
+)
+
+// New creates the route metrics controller. This is the controller
+// that handles all the logic for gathering and exporting
+// metrics related to route resources.
+func New(mgr manager.Manager, namespace string) (controller.Controller, error) {
+	// Create a new cache to watch on Route objects from every namespace.
+	newCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	// Add the cache to the manager so that the cache is started along with the other runnables.
+	mgr.Add(newCache)
+	reconciler := &reconciler{
+		cache:     newCache,
+		Namespace: namespace,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	// add watch for changes in IngressController
+	if err := c.Watch(&source.Kind{Type: &operatorv1.IngressController{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, err
+	}
+	// add watch for changes in Route
+	if err := c.Watch(source.NewKindWithCache(&routev1.Route{}, newCache),
+		handler.EnqueueRequestsFromMapFunc(reconciler.routeToIngressController)); err != nil {
+		return nil, err
+	}
+	return c, nil
+}
+
+// routeToIngressController creates a reconcile.Request for all the Ingress Controllers related to the Route object.
+func (r *reconciler) routeToIngressController(obj client.Object) []reconcile.Request {
+	var requests []reconcile.Request
+	// Cast the received object into Route object.
+	routeObject := obj.(*routev1.Route)
+
+	// Iterate through the related Ingress Controllers.
+	for _, ic := range routeObject.Status.Ingress {
+		log.Info("queueing ingresscontroller", "name", ic.RouterName)
+		// Create a reconcile.Request for the Ingress Controller.
+		request := reconcile.Request{
+			NamespacedName: types.NamespacedName{
+				Name:      ic.RouterName,
+				Namespace: r.Namespace,
+			},
+		}
+		requests = append(requests, request)
+	}
+	return requests
+}
+
+// reconciler handles the actual ingresscontroller reconciliation logic in response to events.
+type reconciler struct {
+	cache     cache.Cache
+	Namespace string
+}
+
+// Reconcile expects request to refer to a Ingress Controller resource, and will do all the work to gather metrics related to
+// the resource.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	log.Info("reconciling", "request", request)
+
+	// Fetch the Ingress Controller object.
+	ingressController := &operatorv1.IngressController{}
+	if err := r.cache.Get(ctx, request.NamespacedName, ingressController); err != nil {
```

Comment:
I'm confused about what exactly are we looking up here.  According to https://github.com/openshift/cluster-ingress-operator/pull/779/files#diff-e0427c17d84da0fa94220e4afa2afde7d3a5d83894c379b9470ffc0fa9766c82R79-R80  it is an arbitrary router name in the namespace of the of the Ingress?  How does that translate to an ingress controller?

Reply:
@Miciah suggested in this comment thread (https://docs.google.com/document/d/1Ojlpm2OTmGPY-5aHcP80HF7deBf1DRWARLrsSkrWneA/edit?disco=AAAAYKzJixo) regarding getting the details from the `Status.Ingress[*].RouterName`. The `RouterName` is obtained from the env var `ROUTER_SERVICE_NAME` which is set to the corresponding Ingress Controller Name here: https://github.com/openshift/cluster-ingress-operator/blob/14f40c789e4e851aae2d812ea8e6fc1e70d9977f/pkg/operator/controller/ingress/deployment.go#L418

Reply:
There comes a problem when multiiple ingress controllers have no labels, because then the route is "Admitted" to all the ingress controllers.  Here is the status for a route with no matching labels in a cluster with two ingress controllers that also don't have selectors.  This means that all the unmarked ingress controllers will be reconciled each time a Route without a label is added.

status:
  ingress:
  - conditions:
    - lastTransitionTime: "2022-09-15T20:50:46Z"
      status: "True"
      type: Admitted
    host: test2-test.apps.ci-ln-zw5f2kk-76ef8.origin-ci-int-aws.dev.rhcloud.com
    routerCanonicalHostname: router-ic2.test2.example.com
    **routerName: ic2**
    wildcardPolicy: None
  - conditions:
    - lastTransitionTime: "2022-09-15T21:06:48Z"
      status: "True"
      type: Admitted
    host: test2-test.apps.ci-ln-zw5f2kk-76ef8.origin-ci-int-aws.dev.rhcloud.com
    routerCanonicalHostname: router-default.apps.ci-ln-zw5f2kk-76ef8.origin-ci-int-aws.dev.rhcloud.com
    **routerName: default**
    wildcardPolicy: None

Reply:
> There comes a problem when multiiple ingress controllers have no labels, because then the route is "Admitted" to all the ingress controllers. Here is the status for a route with no matching labels in a cluster with two ingress controllers that also don't have selectors. This means that all the unmarked ingress controllers will be reconciled each time a Route without a label is added.

As the route is "Admitted" by multiple Ingress Controllers we do need to reconcile for each of them. Otherwise, it will not be exporting the correct metric information for them.

> Also, as your code is written, I think it means that the metric will be deleted from all the unmarked controllers too?

If a route is "Admitted" by multiple Ingress Controllers, it'll be counted for all of them during the calculation of the metrics value. Deletion of a label value of the metric only happens during the deletion event of an Ingress Controller. There's no deletion of metric values any other place.